src/sfma/models/marginal.py

```diff
@@ -117,13 +125,27 @@ def gradient(self, x: ndarray, data: Data) -> ndarray:
         dlerf = np.zeros(z.shape)
         dlerf[index] = -2 * z[index] - 1 / z[index]
         dlerf[~index] = -2 * np.exp(-z[~index]**2) / erfc(z[~index]) / np.sqrt(np.pi)
-        grad = np.zeros(beta.size + 2)
+        grad = np.zeros((beta.size + 2, data.obs.shape[0]))
+
+        grad[:beta.size, ] = x.T * (dlerf*np.sqrt(v_ie/(v_roe*v))/np.sqrt(2) - r/v)
+        grad[beta.size, ] = 0.5*(-r**2/v**2 + 1/v + dlerf*r*np.sqrt(v_ie/(v_roe*v))/np.sqrt(2)*(1/v_roe + 1/v))
+        grad[-1, ] = 0.5*(-r**2/v**2 + 1/v - dlerf*r*np.sqrt(v_ie/(v_roe*v))/np.sqrt(2)*(1/v_ie - 1/v))
+
+        return grad
+
+    def gradient(self, x: ndarray, data: Data) -> ndarray:
+        """
+        Computes the gradient.
 
-        grad[:beta.size] += x.T.dot(dlerf*np.sqrt(v_ie/(v_roe*v))/np.sqrt(2) - r/v)
-        grad[beta.size] += 0.5*np.sum(-r**2/v**2 + 1/v + dlerf*r*np.sqrt(v_ie/(v_roe*v))/np.sqrt(2)*
-                                      (1/v_roe + 1/v))
-        grad[-1] += 0.5*np.sum(-r**2/v**2 + 1/v - dlerf*r*np.sqrt(v_ie/(v_roe*v))/np.sqrt(2)*
-                               (1/v_ie - 1/v))
+        :param x:
+        :param data:
+        :param w: optional weights
+        :return:
+        """
+        grad = self._gradient(x=x, data=data)
+        if self.w is not None:
+            grad = self.w * grad
+        grad = np.sum(grad, axis=1)
 
         # Take the average because the objective
         # is the mean rather than the sum
```

Comment:
!Inconsistency with objective, here gradient is the mean, but the objective function is the sum. Might be the reason why the algorithm have a hard time to converge :)